From 37fe0d8e11c987608517b851d27e9883e349fd72 Mon Sep 17 00:00:00 2001
From: Viktor Ashirov <vashirov@redhat.com>
Date: Thu, 4 Dec 2025 22:09:13 +0100
Subject: [PATCH] Issue 7121 - LeakSanitizer: various leaks during replication
 (#7122)

1. CSN Leaks

In `resolve_attribute_state_deleted_to_present()` we set CSN set pointer
to NULL without freeing the allocated memory.
In `valueset_remove_valuearray()` we overwrite `csnset` pointer without
freeing the existing `csnset`.

2. Leak in replica_add_session_abort_control()

Control's OID and value are allocated but never freed after
`slapi_pblock_set`, which duplicates the control. Added cleanup to free
`ctrl.ldctl_oid` and `ctrl.ldctl_value.bv_val` after `slapi_pblock_set`.

3. LDAP controls leak

`ldap_parse_result` allocates controls that are not being freed when not
transferred to caller or on error paths.  Free `loc_returned_controls`
in cleanup section and NULL the pointer after the transfer.

`returned_controls` allocated in `conn_read_result_ex` are used to check
for abort session control, but never freed before the next loop
iteration.

Fixes: https://github.com/389ds/389-ds-base/issues/7121

Reviewed by: @progier389, @tbordaz (Thanks!)
---
 ldap/servers/plugins/replication/repl5_connection.c   | 4 ++++
 ldap/servers/plugins/replication/repl5_inc_protocol.c | 4 ++++
 ldap/servers/plugins/replication/repl5_replica.c      | 3 +++
 ldap/servers/slapd/entrywsi.c                         | 2 +-
 ldap/servers/slapd/valueset.c                         | 2 ++
 5 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/ldap/servers/plugins/replication/repl5_connection.c b/ldap/servers/plugins/replication/repl5_connection.c
index 690241ebd..aa2a137f7 100644
--- a/ldap/servers/plugins/replication/repl5_connection.c
+++ b/ldap/servers/plugins/replication/repl5_connection.c
@@ -504,6 +504,7 @@ conn_read_result_ex(Repl_Connection *conn, char **retoidp, struct berval **retda
         {
             if (NULL != returned_controls) {
                 *returned_controls = loc_returned_controls;
+                loc_returned_controls = NULL; /* ownership transferred */
             }
             if (LDAP_SUCCESS != rc) {
                 conn->last_ldap_error = rc;
@@ -517,6 +518,9 @@ conn_read_result_ex(Repl_Connection *conn, char **retoidp, struct berval **retda
         slapi_ch_free_string(&errmsg);
         slapi_ch_free_string(&matched);
         charray_free(referrals);
+        if (loc_returned_controls) {
+            ldap_controls_free(loc_returned_controls);
+        }
     }
     if (res)
         ldap_msgfree(res);
diff --git a/ldap/servers/plugins/replication/repl5_inc_protocol.c b/ldap/servers/plugins/replication/repl5_inc_protocol.c
index 41a20c370..14b35766f 100644
--- a/ldap/servers/plugins/replication/repl5_inc_protocol.c
+++ b/ldap/servers/plugins/replication/repl5_inc_protocol.c
@@ -371,6 +371,10 @@ repl5_inc_result_threadmain(void *param)
         if (op) {
             repl5_inc_op_free(op);
         }
+        if (returned_controls) {
+            ldap_controls_free(returned_controls);
+            returned_controls = NULL;
+        }
     }
     slapi_log_err(SLAPI_LOG_REPL, repl_plugin_name, "repl5_inc_result_threadmain exiting\n");
 }
diff --git a/ldap/servers/plugins/replication/repl5_replica.c b/ldap/servers/plugins/replication/repl5_replica.c
index a97c807e9..5180d7ca0 100644
--- a/ldap/servers/plugins/replication/repl5_replica.c
+++ b/ldap/servers/plugins/replication/repl5_replica.c
@@ -4288,6 +4288,9 @@ replica_add_session_abort_control(Slapi_PBlock *pb)
     bvp->bv_val = NULL;
     ber_bvfree(bvp);
     slapi_pblock_set(pb, SLAPI_ADD_RESCONTROL, &ctrl);
+    /* Free the control since slapi_pblock_set duplicates it */
+    slapi_ch_free_string(&ctrl.ldctl_oid);
+    slapi_ch_free_string(&ctrl.ldctl_value.bv_val);
 
     slapi_log_err(SLAPI_LOG_REPL, repl_plugin_name,
                   "add_session_abort_control - abort control successfully added to result\n");
diff --git a/ldap/servers/slapd/entrywsi.c b/ldap/servers/slapd/entrywsi.c
index dbf8a85c8..e1bdc1bab 100644
--- a/ldap/servers/slapd/entrywsi.c
+++ b/ldap/servers/slapd/entrywsi.c
@@ -1185,8 +1185,8 @@ resolve_attribute_state_deleted_to_present(Slapi_Entry *e, Slapi_Attr *a, Slapi_
             if ((csn_compare(vucsn, deletedcsn) >= 0) ||
                 value_distinguished_at_csn(e, a, valuestoupdate[i], deletedcsn)) {
                 entry_deleted_value_to_present_value(a, valuestoupdate[i]);
+                csnset_free(&valuestoupdate[i]->v_csnset);
             }
-            valuestoupdate[i]->v_csnset = NULL;
         }
     }
 }
diff --git a/ldap/servers/slapd/valueset.c b/ldap/servers/slapd/valueset.c
index a66ec2f46..23f9e7007 100644
--- a/ldap/servers/slapd/valueset.c
+++ b/ldap/servers/slapd/valueset.c
@@ -1324,6 +1324,7 @@ valueset_remove_valuearray(Slapi_ValueSet *vs, const Slapi_Attr *a, Slapi_Value
                     if (found->v_csnset &&
                         (flags & (SLAPI_VALUE_FLAG_PRESERVECSNSET |
                                   SLAPI_VALUE_FLAG_USENEWVALUE))) {
+                        csnset_free(&valuestodelete[i]->v_csnset);
                         valuestodelete[i]->v_csnset = csnset_dup(found->v_csnset);
                     }
                     if (flags & SLAPI_VALUE_FLAG_USENEWVALUE) {
@@ -1334,6 +1335,7 @@ valueset_remove_valuearray(Slapi_ValueSet *vs, const Slapi_Attr *a, Slapi_Value
                     }
                 } else {
                     if (flags & SLAPI_VALUE_FLAG_PRESERVECSNSET) {
+                        csnset_free(&valuestodelete[i]->v_csnset);
                         valuestodelete[i]->v_csnset = found->v_csnset;
                         found->v_csnset = NULL;
                     }
-- 
2.52.0

